topdown/num_bytes.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2016 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package topdown
+
+import (
+	"fmt"
+	"math/big"
+	"strconv"
+	"strings"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/topdown/builtins"
+)
+
+type multiplier int
+
+const (
+	none multiplier = 1
+	kb              = 1000
+	ki              = 1024
+	mb              = kb * 1000
+	mi              = ki * 1024
+	gb              = mb * 1000
+	gi              = mi * 1024
+	tb              = gb * 1000
+	ti              = gi * 1024
+)
+
+// The rune values for 0..9 as well as the period symbol (for parsing floats)
+var numRunes = []rune("0123456789.")
+
+func parseNumBytesError(msg string) error {
+	return fmt.Errorf("%s error: %s", ast.ParseNumBytes.Name, msg)
+}
+
+func errUnitNotRecognized(unit string) error {
+	return parseNumBytesError(fmt.Sprintf("byte unit %s not recognized", unit))
+}
+
+var (
+	errNoAmount = parseNumBytesError("no byte amount provided")
+	errIntConv  = parseNumBytesError("could not parse byte amount to integer")
+)
+
+func builtinNumBytes(a ast.Value) (ast.Value, error) {
+	var m multiplier
+
+	raw, err := builtins.StringOperand(a, 1)
+	if err != nil {
+		return nil, err
+	}
+
+	s := formatString(raw.String())
+
+	numStr, unitStr := extractNumAndUnit(s)
+
+	if numStr == "" {
+		return nil, errNoAmount
+	}
+
+	switch unitStr {
+	case "":
+		m = none
+	case "kb":
+		m = kb
+	case "kib":
+		m = ki
+	case "mb":
+		m = mb
+	case "mib":
+		m = mi
+	case "gb":
+		m = gb
+	case "gib":
+		m = gi
+	case "tb":
+		m = tb
+	case "tib":
+		m = ti
+	default:
+		return nil, errUnitNotRecognized(unitStr)
+	}
+
+	num, err := strconv.Atoi(numStr)
+	if err != nil {
+		return nil, errIntConv
+	}
+
+	total := num * int(m)
+
+	return builtins.IntToNumber(big.NewInt(int64(total))), nil
+}
+
+// Makes the string lower case and removes spaces and quotation marks
+func formatString(s string) string {
+	lower := strings.ToLower(s)
+	stripped := strings.Replace(lower, " ", "", -1)
```

Comment:
I think this ought to be handled in extractNumAndUnit so that spaces between number runes and units aren't accepted. It would be good to update the test suite to cover these (e.g., "1 1 gb" => error, "1 1" => error, "1 g b" => error, etc.)

Reply:
So, I update it to disallow *all* spaces in resource strings. I'm honestly not quite sure how strict we should be about that. I can change that if you like.

Reply:
That's also fine.